### PR TITLE
librbd: default features should be negotiated with the OSD

### PR DESCRIPTION
**Pull request by trociny**
Fixes: http://tracker.ceph.com/issues/17010

### CODE REVIEW

**Comment by trociny** [comment]
@dillaman Not sure about util::get_rbd_default_features(). I mostly need it for internal.cc, but it will not work for a new pool, because of RBD_DIRECTORY not existing yet.

File: src/librbd/Utils.cc [hunk 1]
```diff
@@ -89,7 +91,7 @@ uint64_t parse_rbd_default_features(CephContext* cct)
         value += conf_vals[feature];
       } else {
         ret = -EINVAL;
-        ldout(cct, 1) << "Warning: unknown rbd feature " << feature << dendl;
+        lderr(cct) << "unknown rbd feature " << feature << dendl;
```
Comment (dillaman) [inline src/librbd/Utils.cc]: Note that this falls back to the defaults on -EINVAL below. That was the rationale for the warning since execution will continue.

**Comment by dillaman** [comment]
@trociny I was thinking that this negotiation would only occur if using the default image features (i.e. the user didn't supply any features when creating the image). If the user supplied a feature not supported by the server, I'd say let it fail. 

In terms of the existence of rbd_directory, if this check could be added to the create state machine after a (modified) validate pool step where you would create the directory if it didn't exist (if pool validation was disabled).

File: src/librbd/Utils.cc (hunk 1 above)
Comment (trociny) [inline src/librbd/Utils.cc]: Yes, I am aware that execution will continue, but I think it is better to display it as an error message to the user (though not fatal). Without this and with `rbd debug = 0` by default, the user will not see e.g. that she has a typo in ceph.conf.

Comment (dillaman) [inline src/librbd/Utils.cc]: Makes sense. Perhaps reword it to be "ignoring unknown feature ..."

File: src/librbd/image/CreateRequest.cc [hunk 2]
```diff
@@ -381,6 +382,48 @@ Context *CreateRequest<I>::handle_add_image_to_directory(int *result) {
...
+template<typename I>
+Context *CreateRequest<I>::handle_negotiate_features(int *result) {
+  ldout(m_cct, 20) << __func__ << ": r=" << *result << dendl;
+
+  uint64_t all_features;
+  if (*result == 0) {
+    bufferlist::iterator it = m_outbl.begin();
+    *result = cls_client::get_all_features_finish(&it, &all_features);
+  }
+  if (*result < 0) {
```
Comment (dillaman) [inline src/librbd/image/CreateRequest.cc]: Minor: probably good to check for ```-EOPNOTSUPP``` and skip the negotiation if it fails.

File: src/librbd/image/CreateRequest.cc [hunk 3]
```diff
@@ -381,6 +382,48 @@ Context *CreateRequest<I>::handle_add_image_to_directory(int *result) {
...
+    ldout(m_cct, 10) << "error retrieving server supported features set: "
+                     << cpp_strerror(*result) << dendl;
+  } else if (all_features < m_features) {
```
Comment (dillaman) [inline src/librbd/image/CreateRequest.cc]: Minor: ```if ((m_features & all_features) != m_features)``` is probably clearer

**Comment by dillaman** [comment]
@trociny For the create() w/ no features case, you could just have the it call the create method that takes the ImageOptions instead of that intermediate version for stripe units.

**Comment by trociny** [comment]
@dillaman good idea! Thanks for your comments.

File: src/librbd/image/CreateRequest.cc (hunk 2 above)
Comment (trociny) [inline src/librbd/image/CreateRequest.cc]: @dillaman Not sure about this. It is only a debug message, so it looks like it does not hurt to display it even for EOPNOTSUPP case.

**Comment by trociny** [comment]
@dillaman updated

**Comment by dillaman** [comment]
lgtm
